crates/executor/src/simulate.rs

```diff
@@ -126,11 +133,31 @@ pub fn trace(
 ) -> Result<Vec<(TransactionHash, TransactionTrace)>, TransactionExecutionError> {
     let (mut state, block_context) = execution_state.starknet_state()?;
 
-    let cached = { cache.0.lock().unwrap().cache_get(&block_hash).cloned() };
-    if let Some(cached) = cached {
-        tracing::trace!(block=%block_hash, "trace cache hit");
-        return Ok(cached);
-    }
+    let sender = {
+        let mut cache = cache.0.lock().unwrap();
+        match cache.cache_get(&block_hash) {
+            Some(CacheItem::Cached(cached)) => {
+                tracing::trace!(block=%block_hash, "trace cache hit");
+                return Ok(cached.clone());
+            }
+            Some(CacheItem::Inflight(receiver)) => {
+                tracing::trace!(block=%block_hash, "trace already inflight");
+                let mut receiver = receiver.subscribe();
+                drop(cache);
```

Comment:
Ugh. This is pretty ugly here...

Reply:
Yeah I wasn't quite sure how to let the guard do the work. Suggestions welcome 🤷